### DISCUSSION TITLE
Precompiled file to load

### DISCUSSION THREAD

**Discussion by zandercodes** [eh-steve/goloader#16] [Ideas]
Can you add Load from Precompiled file like .o?

**Comment by eh-steve** [comment]
What would this file contain?

**Reply by zandercodes** [reply]
It is yes a file compile as you described at the Example is described. That happens currently only at runtime.
Wouldn't it be possible to precompile a file so that you don't need `jit.BuildGo...` anymore, but can access the finished file that is in the system, similar to the go internal plugin system?

**Reply by eh-steve** [reply]
I’d say .o isn’t a great extension, since this would be a collection of archives (.a) each of which could contain one or more .o files.

This is potentially possible, but there are 2 problems:

- The archives are not cross-OS and architecture compatible (same as Go binaries), unlike sources, so you’d need a different bundle of archives for each OS/arch (macho/elf/PE * amd64/arm64)
- Currently we only build and load the packages where symbols are missing and not already included in the host binary, unlike the Go plugin system which includes the whole runtime etc every time, even if these parts are available in the host. We can only know which symbols will be missing for a given host binary at runtime. We also only include packages whose symbols are reachable from the target package (deadcode is eliminated). This is one of the reasons why loading a small JIT module in a large host binary can be faster/lighter than using a plugin.

I could add a build util which bundles all possible packages (including deadcode elimination) into a single archive which can be loaded on any host binary of that OS/arch but these archives would be quite big/slow to load. Potentially, if packages were indexed by name in this archive, we could only lazy-load packages we need so preserving a bit of speed… 

I could also make a util which builds an archive specifically designed for a certain host binary, and therefore could omit certain symbols/packages on the expectation that they’d be available in the host binary.

Which of these 2 options would be more useful to you?

**Reply by eh-steve** [reply]
I think I will implement this at some point, as it will allow goloader to be closer to a drop-in alternative to `plugin`
